### PR TITLE
Add ImplementationAttributes to IMethodSymbol

### LINKED ISSUE #7282: Expose IMethodSymbol.ImplementationAttributes
Currently we expose `IMethodSymbol.GetDllImportData` which contains the information about the P/Invoke attributes. Both C# and VB compilers populate this data during well known attribute decoding. This is very helpful information for analyzers looking at P/Invoke methods.

We also compute and store the [ImplementationAttributes](http://source.roslyn.io/Microsoft.CodeAnalysis.CSharp/R/37df62720b80ad32.html) that apply to the method during attribute decoding, but it is internal - I already found FxCop rules that need information such as PreserveSig data, and we should expose this to enable public analysis of P/invoke methods.

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #7282

### CODE REVIEW

File: src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationAbstractMethodSymbol.cs
Comment (Youssef1313) [inline src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationAbstractMethodSymbol.cs:35]: This *might* be incorrect. Should I attempt to retrieve it through the `attributes` passed to the ctor of this abstract class? Hope someone can confirm.

File: src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationMethodSymbol.cs
Comment (Youssef1313) [inline src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationMethodSymbol.cs:102]: This assumes `MethodImplAttributes.IL`. If this is wrong, it could be related to my comment https://github.com/dotnet/roslyn/pull/50560/files#r559016146

**Comment by jcouv** [comment]
Tagging @dotnet/roslyn-compiler: I think it's sensible to make the `ImplementationAttributes` API public. @mavasani had identified some scenarios where that's useful. Any concerns?
Manish, can you confirm those scenarios are still applicable (we didn't already find another solution)?

@Youssef1313 Assuming we can get agreement on design change, a PR that exposes new public APIs needs direct test coverage for those APIs.

**Comment by 333fred** [comment]
> Tagging @dotnet/roslyn-compiler: I think it's sensible to make the ImplementationAttributes API public. @mavasani had identified some scenarios where that's useful. Any concerns?

My main question is: why does this need special handling? It's an attribute, right? You should be able get it through the regular API?

**Comment by 333fred** [comment]
(Didn't mean to close, sorry)

**Comment by jcouv** [comment]
@333fred Those are for metadata flags. You can put set them via attributes in source, but they are not emitted as attributes.
See [PopulateMethodTableRows](https://github.com/dotnet/roslyn/blob/bf6ad0f9f81eca3172e428e46301fbc35b71f2ba/src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs#L2568-L2575) and `ImplementationAttributes` in PE symbols.
Here's a [sharplab example](https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgJgIwFgBQyAMABMugHQDiEA9gEYCGEYAXrcGJQHYDcO+RpASgFd2rALYBTEgGFKogA6RxAJwDKygG5gAxuIDO3bDwDMRVAQCqrBsACeOAN44CzgNoBZccAAWlGAEl5CABBYGAlMGpBYHEACg9vXwC5CAB5OVYOXRIAOUo/dgZ2MHYAcwBKAF0CJ2dkE2IANj5CMk8penF2GFolbNoJGPaITu6lAi16MpqCR2xneaIAdnH6EgAVShUw4pKYspJ+cWTaHRiAImJyKjoGZgz2EjOoAjOz/emFz8/D49OzoZGPSeLzeBnmAF8cOCgA) to illustrate.

**Comment by 333fred** [comment]
Thanks for the clarification Julien. So, the main concern here is symbols from metadata, since symbols from source could retrieve the info via the existing attributes APIs?

**Comment by jcouv** [comment]
There seems to be some additional value, even aside from metadata symbols.
Some methods get implementation flags for other reasons than having an `MethodImplAttribute` on them. See `SourceMethodSymbolWithAttributes.ImplementationAttributes`, `SynthesizedInstanceConstructor` and `SynthesizedEventAccessorSymbol` for example.

File: src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs
Comment (333fred) [inline src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs:267]: Needs documentation comments.

File: src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationAbstractMethodSymbol.cs
Comment (333fred) [reply src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationAbstractMethodSymbol.cs:35]: This stuff is used by the IDE when generating methods. I don't believe they're going to care about this API, so I don't think it actually matters. @jasonmalinowski to confirm what they'd like the impl strategy to be though.

**Comment by 333fred** [comment]
Done review pass (commit 2)

File: src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs
Comment (Youssef1313) [inline src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs:48]: @jcouv Before I attempt to write more tests, I want to confirm that I'm doing it the correct way. I feel that I'm missing something. Both IL and Managed are defined in the enum as 0, which confused me a little bit.

Comment (333fred) [reply src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs:48]: If it's a managed method, it's implemented in MSIL. It's certainly not the greatest thing to have 2 labels, but it's out of our control :). No need to test equality to both. Also, please use `Assert.Equals`, not `Assert.True(==)`.

Comment (Youssef1313) [reply src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs]: @333fred Thanks. I've updated it to use `Assert.Equal`. Do you want me to keep updating tests in this file? or creating a brand new test file for testing this API?

Comment (333fred) [reply src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs]: This file is fine.

File: src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs:46]: Since `Managed` and `IL` are the same value, we don't need to compare to both. Just one is fine.

File: src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs:947]: We need to verify these from both source and metadata. Instead of getting the symbol from the compilation like this, create a helper that, given a ModuleSymbol, finds and verifies the symbols. You can then pass that as the `symbolValidator` and `sourceSymbolValidator` parameters to `CompileAndVerify`. Same for the test of these added tests.

File: src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs
Comment (jasonmalinowski) [inline src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs:273]: Is there any concern here about using a type from the framework that Roslyn _targets_? i.e. if coreclr adds a new value and the compiler is compiling a project we might see values that aren't in the enum version for what we're targeting. This is _fine_ in that enums don't have any validation but not sure if it's worth a comment this is a potential scenario that could happen...?

If we have a precedent for this sort of API I defer to whatever we did of course.

File: src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationAbstractMethodSymbol.cs
Comment (jasonmalinowski) [reply src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationAbstractMethodSymbol.cs:35]: Agreed, we're be unlikely to care. These symbols are created when we we're using some of our code generation APIs that generate syntax from IMethodSymbols. Unless that code is also being updated to actually _generate_ this method, nothing is going to look at any implementation here.

So just to clarify, what's the behavior of symbols you get from the compiler?

**From Metadata:** ImplementationAttributes is correct, GetAttributes returns empty (since it's baked into the flags...?)
**From Source:** GetAttributes returns symbols, ImplementationAttributes is ...?

File: src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs
Comment (333fred) [reply src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs:273]: > Is there any concern here about using a type from the framework that Roslyn targets?

We do this in a few other places as well. Personally, I don't see a huge issue with it: it doesn't feel all that different from compiling against an older version of the Roslyn API and seeing unexpected values in the `SyntaxKind` enum.

Comment (jasonmalinowski) [reply src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs:273]: Sounds good to me then.

File: src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs
Comment (Youssef1313) [reply src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs]: @333fred I updated one test. Can you check whether the change is correct? If correct, I'll update the rest of them. Otheriwse, I'll need some guidance. Thanks.

File: src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationAbstractMethodSymbol.cs
Comment (333fred) [reply src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationAbstractMethodSymbol.cs:35]: > So just to clarify, what's the behavior of symbols you get from the compiler?

@jasonmalinowski `ImplementationAttributes` will be correct in all cases. In source cases, you'd get the attributes in `GetAttributes` as well.

File: src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs
Comment (333fred) [reply src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs]: @Youssef1313 looks correct. The only comment is that `CompileAndVerify` also ensures there are no diagnostics, so you can remove the `CreateCompilation` and `VerifyDiagnostics` pair.

**Comment by Youssef1313** [comment]
@333fred Can you assist with the test failures?

```
[xUnit.net 00:00:19.07]     Microsoft.CodeAnalysis.CSharp.UnitTests.Symbols.SymbolEqualityTests.TestImplementationAttributes_Native [FAIL]
  Failed Microsoft.CodeAnalysis.CSharp.UnitTests.Symbols.SymbolEqualityTests.TestImplementationAttributes_Native [8 s]
  Error Message:
   System.Exception : Verification failed
---- Roslyn.Test.Utilities.Desktop.RuntimePeVerifyException :
PeVerify failed for assembly 'C:\Users\Youssef\AppData\Local\Temp\RoslynTests':
Type load failed.
```

```
[xUnit.net 00:00:20.29]     Microsoft.CodeAnalysis.CSharp.UnitTests.Symbols.SymbolEqualityTests.TestImplementationAttributes_PreserveSigAndRuntime [FAIL]
  Failed Microsoft.CodeAnalysis.CSharp.UnitTests.Symbols.SymbolEqualityTests.TestImplementationAttributes_PreserveSigAndRuntime [158 ms]
  Error Message:
   System.Exception : Verification failed
---- Roslyn.Test.Utilities.Desktop.RuntimePeVerifyException :
PeVerify failed for assembly 'C:\Users\Youssef\AppData\Local\Temp\RoslynTests':
Error: Method marked Abstract, Runtime, InternalCall or Imported must have zero RVA, and vice versa.
Type load failed.
```

File: src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs
Comment (333fred) [inline src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs:1083]: @Youssef1313 PEVerify probably doesn't like that you've declared a method with a managed body and marked it `Native`. Probably should make it extern.

Comment (333fred) [reply src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs]: Looking at another test in our codebase, you can just go ahead and skip PEVerify. Use `verify: Verification.Skipped` in the `CompileAndVerify` options.

File: src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs:47]: >Assert.Equal(MethodImplAttributes.IL, nonNullPlus.ImplementationAttributes); [](start = 12, length = 76)

Could you please elaborate what are we trying to test here?

File: src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs:924]: >TestImplementationAttributes_Inlining [](start = 20, length = 37)

I do not expect the new API to have any impact on symbol equality. Any changes in this file are unexpected to me.

Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs:924]: >TestImplementationAttributes_Inlining [](start = 20, length = 37)

It looks like we are lacking tests for VB.

File: src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs:271]: >These attributes are not always emitted as attributes. [](start = 12, length = 54)

These are not attributes at all. If this is what we want to say, we should say that.

File: src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs:273]: >ImplementationAttributes [](start = 29, length = 24)

Should we name the property "MethodImplAttributes"?

**Comment by AlekseyTs** [comment]
Done with review pass (commit 13)

File: src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs
Comment (Youssef1313) [reply src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs:924]: @AlekseyTs Any recommendation where the tests should go?

Comment (Youssef1313) [reply src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs:924]: @AlekseyTs I'll add them.

File: src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs
Comment (Youssef1313) [reply src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs:271]: @AlekseyTs Any recommendation for a good doc comment?

File: src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs
Comment (Youssef1313) [reply src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs:273]: @AlekseyTs It already exists with "ImplementationAttributes" name over the codebase.

File: src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Symbol/Symbols/SymbolEqualityTests.cs:924]: > Any recommendation where the tests should go?

I think a new file under src\Compilers\CSharp\Test\Symbol\Symbols\ would be fine, or a "good" existing file there. Same for VB.

---
In reply to: [569592234](https://github.com/dotnet/roslyn/pull/50560#discussion_r569592234) [](ancestors = 569592234)

File: src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs:273]: > It already exists with "ImplementationAttributes" name over the codebase.

That is an internal API and we aren't that picky about names of internal APIs. There are no concerns about if the name is going to confuse external consumers, etc. For example, it looks like reflection API is using ```GetMethodImplementationFlags``` name. Since we are returning the same information and using the same type, it is probably good to use terms already familiar to consumers. "MethodImplementationFlags", for example.

---
In reply to: [569594171](https://github.com/dotnet/roslyn/pull/50560#discussion_r569594171) [](ancestors = 569594171)

File: src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs
Comment (AlekseyTs) [reply src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs:271]: > Any recommendation for a good doc comment?

I would remove the remark section completely. And change description to: "Returns the implementation flags for the given method." Any additional information can be found in documentation for the return type, etc. If we want to be very specific, in addition we could say what does this value map to in metadata (table, field). But I don't think this is necessary.

---
In reply to: [569592792](https://github.com/dotnet/roslyn/pull/50560#discussion_r569592792) [](ancestors = 569592792)

**Comment by jcouv** [comment]
@Youssef1313 I marked the PR as draft while awaiting revisions. Please ping us when ready for another look. Thanks

**Comment by Youssef1313** [comment]
@jcouv Thanks for the ping. I addressed @AlekseyTs's feedback. But not sure why some tests are failing. Failures happens for AggressiveInlining only.

File: src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/MethodSymbol.cs:133]: >MethodImplementationFlags [](start = 65, length = 25)

In my opinion this rename is unnecessary.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs:3125]: >MethodImplementationFlags [](start = 145, length = 25)

The rename looks suspicious.

File: src/Compilers/Test/Utilities/VisualBasic/MockSymbols.vb
Comment (AlekseyTs) [inline src/Compilers/Test/Utilities/VisualBasic/MockSymbols.vb:461]: >MethodImplementationFlags [](start = 39, length = 25)

Similar to C# I think this rename is unnecessary.

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/PublicAPI.Unshipped.txt:17]: >ImplementationAttributes [](start = 37, length = 24)

Is there an API like that.

File: src/Compilers/CSharp/Test/Symbol/Symbols/MethodImplementationFlagsTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/MethodImplementationFlagsTests.cs:40]: >aggressiveInliningMethod.MethodImplementationFlags [](start = 70, length = 50)

It looks like this test doesn't actually test the added public API. Rename of the internal API made it hard to spot. I assume the other tests in this file have the same issue.

File: src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/MethodImplementationFlagsTests.vb
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/MethodImplementationFlagsTests.vb:36]: >aggressiveInliningMethod.MethodImplementationFlags [](start = 113, length = 50)

Same comment  as for C# tests.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 17)

File: src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationMethodSymbol.cs
Comment (Youssef1313) [inline src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationMethodSymbol.cs:102]:
```suggestion
        public override System.Reflection.MethodImplAttributes MethodImplementationFlags => default;
```

File: src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationConstructedMethodSymbol.cs
Comment (Youssef1313) [inline src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationConstructedMethodSymbol.cs:74]:
```suggestion
        public override System.Reflection.MethodImplAttributes MethodImplementationFlags => _constructedFrom.MethodImplementationFlags;
```

File: src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationAbstractMethodSymbol.cs
Comment (Youssef1313) [inline src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationAbstractMethodSymbol.cs:35]:
```suggestion
        public abstract System.Reflection.MethodImplAttributes MethodImplementationFlags { get; }
```

File: src/Features/Core/Portable/MetadataAsSource/AbstractMetadataAsSourceService.WrappedMethodSymbol.cs
Comment (Youssef1313) [inline src/Features/Core/Portable/MetadataAsSource/AbstractMetadataAsSourceService.WrappedMethodSymbol.cs:36]:
```suggestion
            public System.Reflection.MethodImplAttributes MethodImplementationFlags => _symbol.MethodImplementationFlags;
```

**Review by 333fred — APPROVED** [review]
LGTM (commit 23), assuming tests pass.

File: src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/MethodImplementationFlagsTests.vb
Comment (333fred) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/MethodImplementationFlagsTests.vb:41]:
```suggestion
            CompileAndVerify(src, sourceSymbolValidator:=validator, symbolValidator:=validator, useLatestFramework:=True)
```

File: src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/MethodImplementationFlagsTests.vb
Comment (333fred) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/MethodImplementationFlagsTests.vb:113]:
```suggestion
            CompileAndVerify(src, sourceSymbolValidator:=validator, symbolValidator:=validator, useLatestFramework:=True)
```

**Review by 333fred — COMMENTED** [review]
These should fix the tests. The issue is that the default VB test framework is too old and doesn't have `AggressiveInlining`, so it doesn't compile.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 23), the CI is failing.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 24)

**Comment by AlekseyTs** [comment]
@Youssef1313 Thanks for the contribution.

**Comment by HelloKitty** [comment]
What version is this shipped in? Currently getting: `System.TypeLoadException: Method 'get_MethodImplementationFlags' in type 'Microsoft.CodeAnalysis.CodeGeneration.CodeGenerationMethodSymbol' from assembly 'Microsoft.CodeAnalysis.Workspaces, Version=3.9.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' does not have an implementation.`

When referencing it was a transient property in a source generator.

**Comment by 333fred** [comment]
Should be 3.10.

**Pull request merged**
